[fd/external] Fix resuming downloads with aria2c

tcely opened the pull request:
Closes #15151

### Description of your *pull request* and other information

<!--

Explanation of your *pull request* in arbitrary form goes here. Please **make sure the description explains the purpose and effect** of your *pull request* and is worded well enough to be understood. Provide as much **context and examples** as possible

-->

Since aria2c was added the `-c` flag has been used.
This doesn't actually work, or make sense, without configuring the stream selector to use `inorder` instead of `default` for the algorithm.

Ideally, `aria2c` would write the needed information for resuming downloads in its own file before exiting. Otherwise, it's more reliable to download everything again.

Change the arguments to have `aria2c` try to resume, then overwrite a partial file if it can't resume, instead of aborting the download.

Arguments:
- `--auto-save-interval=10`: save the control file for the transfer to disk faster than the default.
    Users can change this as they like without impacting the success of the transfer.
- `--remove-control-file`: for resume to work, this must be `false` because `true` removes the control file before attempting to resume the transfer.
    Users can change this, but it's required to be `false` for `continuedl` to work as expected.
- `--allow-overwrite=true`: Without this the transfer is aborted instead of overwriting the file.
    Users should not change this.
- `--always-resume=false`: Without this transfers that cannot be resumed are aborted.
    Users should not change this.
- `--auto-file-renaming=false`: Without this the output file is changed to a filename that doesn't already exist.
    Users should not change this.
- `--force-save=false`: The control file should be removed (not saved) after the transfer is completed.
    Users should not change this.

<!-- Fixes # -->


<details open><summary>Template</summary> <!-- OPEN is intentional -->

<!--

# PLEASE FOLLOW THE GUIDE BELOW

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes `[ ]` relevant to your *pull request* (like [x])
- Use *Preview* tab to see how your *pull request* will actually look like

-->

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [x] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

tcely committed:
Remove continue from aria2c

This doesn't actually work, or make sense, without configuring the stream selector to use `inorder` instead of `default` for the algorithm.

Ideally, `aria2c` would write the needed information for resuming downloads in its own file before exiting.
Otherwise, it's more reliable to download everything again.

tcely committed:
aria2c should overwrite the output when resuming fails

Moved renaming and overwrite arguments up with the rest.
Both cases need these arguments.

Save the control file every 10 seconds.
Resume using the control file, or overwrite if not possible.
Remove the control file when the download is completed.

Print to stderr instead, just as curl does.

tcely committed:
Honor continuedl with aria2c

--remove-control-file[=true|false]

Remove control file before download.
Using with --allow-overwrite=true,
download always starts from scratch.

pukkandan commented on yt_dlp/downloader/external.py:
```diff
@@ -331,7 +335,6 @@ def _make_cmd(self, tmpfilename, info_dict):
             cmd += ['--dir', self._aria2c_filename(dn) + os.path.sep]
         if 'fragments' not in info_dict:
             cmd += ['--out', self._aria2c_filename(os.path.basename(tmpfilename))]
-        cmd += ['--auto-file-renaming=false']
```
This is intentionally kept after `_configuration_args` so that user cannot override it

tcely replied:
Alright. I've adjusted this and the other flags a user should not override.

pukkandan commented on yt_dlp/downloader/external.py:
```diff
@@ -269,6 +269,7 @@ def _make_cmd(self, tmpfilename, info_dict):
 class Aria2cFD(ExternalFD):
     AVAILABLE_OPT = '-v'
     SUPPORTED_PROTOCOLS = ('http', 'https', 'ftp', 'ftps', 'dash_frag_urls', 'm3u8_frag_urls')
+    _CAPTURE_STDERR = False
```
Why?

tcely replied:
No good reason. Removed.

pukkandan commented on yt_dlp/downloader/external.py:
```diff
@@ -293,11 +294,13 @@ def _call_downloader(self, tmpfilename, info_dict):
 
     def _make_cmd(self, tmpfilename, info_dict):
-        cmd = [self.exe, '-c', '--no-conf',
+        cmd = [self.exe, '--no-conf', '--stderr=true',
+               '--auto-save-interval=10', '--allow-overwrite=true', '--always-resume=false',
+               '--auto-file-renaming=false', '--force-save=false',
```
Pls explain why you believe each of these is suitable for us. The PR description does not cover all of them

tcely replied:
I'll add each of the new flags to the description.

tcely committed:
Move required flags below _configuration_args()

- Resume or overwrite the specified file
- Do not rename output to a non-existent file
- Do not abort the transfer when resume fails
- Do not preserve the control file after transfers complete

tcely committed:
Remove stderr changes

pukkandan commented:
> * `--auto-save-interval=10`: save the control file for the transfer to disk faster than the default.
>   Users can change this as they like without impacting the success of the transfer.

But why is the default insufficient?

> * `--remove-control-file`: for resume to work, this must be `false` because `true` removes the control file before attempting to resume the transfer.
>   Users can change this, but it's required to be `false` for `continuedl` to work as expected.

The default is false, no? We can enforce this for `continuedl` only

> * `--force-save=false`: The control file should be removed (not saved) after the transfer is completed.
>   Users should not change this.

This is also default, no? Could you explain why user shouldn't be able to change this?

tcely commented:
> > --auto-save-interval=10

> But why is the default insufficient?

The setting means there is no way to resume unless:

1. the `aria2c` binary exits cleanly OR
2. it has been transferring for longer than the configured value.

The value of '0' removes option 2.
After the transfer has made it past 10, I would like it to be able to resume if something killed the transfer without waiting for a clean shutdown to write a control file.
I wouldn't object to setting it to 60 either, but it should not be zero, if we want a reasonable chance to resume a transfer.

> > `--remove-control-file=false`

> The default is `false`, no? We can enforce this for `continuedl` only

Based on a skim through the source, this default is unclear to me and it's not documented.
Setting it for the user depending on how `continuedl` is configured and allowing them to override it too seems appropriate to me.

> > `--force-save=false`

> This is also default, no? Could you explain why user shouldn't be able to change this?

There is all manner of bad advice about this flag on the Internet. It's intended purpose is to keep the control file around after a BitTorrent transfer because the seeding statistics are stored in that file.

Allowing this to be set to `true` will only lead to extra unnecessary control files. It's not ever going to be useful for `yt_dlp` because there isn't a saved session file being loaded.

tcely committed:
Merge branch 'master' into patch-1

tcely committed:
Merge branch 'master' into patch-1

tcely committed:
Merge branch 'master' into patch-1
